File: 843. Guess the Word/solution_minimax.cc
/**
 * // This is the Master's API interface.
 * // You should not implement it, or speculate about its implementation
 * class Master {
 *   public:
 *     int guess(string word);
 * };
 */
class Solution {
public:
    int distance(string w1, string w2) {
        if (w1.size() != w2.size()) return -2;

        int matches = 0;

        for (int i = 0; i < w1.size(); i++) {
            if (w1[i] == w2[i]) {
                matches++;
            }
        }

        return matches;
    }

    void findSecretWord(vector<string>& words, Master& master) {
        for (int i = 0, match = 0; i < 10 and match < 6; i++) {
            unordered_map<string, int> wordCount;

            for (string w1 : words) {
                for (string w2 : words) {
                    if (distance(w1, w2) == 0) {
                        wordCount[w1]++;
                    }
                }
            }

            pair<string, int> minimax = { words[0], 10000 };

            for (string w : words) {
                if (wordCount[w] <= minimax.second) {
                    minimax = make_pair(w, wordCount[w]);
                }
            }

            int masterMatch = master.guess(minimax.first);

            vector<string> candidateWords = {};

            for (string w : words) {
                if (distance(w, minimax.first) == masterMatch) {
                    candidateWords.push_back(w);
                }
            }
            words = candidateWords;
        }
    }
};

File: 843. Guess the Word/solution_minimax.py
# """
# This is Master's API interface.
# You should not implement it, or speculate about its implementation
# """
# class Master:
#     def guess(self, word: str) -> int:

class Solution:
    def distance(self, word1: str, word2: str) -> int:
        if len(word1) != len(word2):
            return -2
        
        matches = 0
        
        for i in range(len(word1)):
            if word1[i] == word2[i]:
                matches += 1

        return matches
    
    def findSecretWord(self, words: List[str], master: 'Master') -> None:
        for i, match in zip(range(0, 10), range(0, 6)):
            word_cnt = defaultdict(int)

            for w1 in words:
                for w2 in words:
                    if self.distance(w1, w2) == 0:
                        word_cnt[w1] += 1

            min_max = (words[0], 10000)

            for w in words:
                if w in word_cnt and word_cnt[w] <= min_max[1]:
                    min_max = (w, word_cnt[w])
            
            master_match = master.guess(min_max[0])
            candidate_words = []

            for w in words:
                if self.distance(w, min_max[0]) == master_match:
                    candidate_words.append(w)
            
            words = candidate_words
